src/data/mv_polynomial/basic.lean

```diff
@@ -93,15 +93,14 @@ namespace mv_polynomial
 variables {σ : Type*} {a a' a₁ a₂ : R} {e : ℕ} {n m : σ} {s : σ →₀ ℕ}
 
 section comm_semiring
-variables [comm_semiring R] {p q : mv_polynomial σ R}
+variables [comm_semiring R] [comm_semiring S₁] {p q : mv_polynomial σ R}
 
 instance decidable_eq_mv_polynomial [decidable_eq σ] [decidable_eq R] :
   decidable_eq (mv_polynomial σ R) := finsupp.decidable_eq
 instance : comm_semiring (mv_polynomial σ R) := add_monoid_algebra.comm_semiring
 instance : inhabited (mv_polynomial σ R) := ⟨0⟩
-instance : has_scalar R (mv_polynomial σ R) := add_monoid_algebra.has_scalar
-instance : semimodule R (mv_polynomial σ R) := add_monoid_algebra.semimodule
-instance : algebra R (mv_polynomial σ R) := add_monoid_algebra.algebra
+instance [semimodule R S₁] : semimodule R (mv_polynomial σ S₁) := add_monoid_algebra.semimodule
+instance [algebra R S₁] : algebra R (mv_polynomial σ S₁) := add_monoid_algebra.algebra
```

Comment:
Do we need an `is_scalar_tower` as well?

Reply:
`add_monoid_algebra.algebra` doesn't currently accept one - right now, this change is just "support the full range of things that add_monoid_algebra supports".

Reply:
I'm confused about what you mean. Certainly `R -> S1 -> mv_polynomial \s S1` is a scalar tower, right?
Ooh, you mean the proof wouldn't be `add_monoid_algebra.is_scalar_tower`. Fair enough.

I still think we should have this instance. It will be annoying if the algebra instance is automatically found, but the scalar tower isn't.

Reply:
This instance would eliminate some `comap` in my work in progress about finite presentation (just to give an example), so I would be very happy to have it... but I think we really need `is_scalar_tower`.

Reply:
Oh, I didn't understand your question. I thought to meant I needed `[is_scalar_tower ...]`, not `instance : is_scalar_tower`

Reply:
I agree we need this instance, but `add_monoid_algebra.is_scalar_tower` doesn't yet exist!

I think that makes more sense as a follow-up - the `mv_polynomial` is full of `mv_polynomial X R →ₐ[R] _` API that probably would need generalizing to `mv_polynomial X R →ₐ[S] _` before is_scalar_tower becomes useful / needed.

Reply:
```suggestion
instance [algebra R S₁] : algebra R (mv_polynomial σ S₁) := add_monoid_algebra.algebra
/-
TODO
instance : is_scalar_tower R S₁ (mv_polynomial σ S₁) := sorry
-/
```

Reply:
Sure, will add that once the build is passing, probably with `add_monoid_algebra.is_scalar_tower` instead of `sorry`

Reply:
Merci!

Reply:
```suggestion
instance [algebra R S₁] : algebra R (mv_polynomial σ S₁) := add_monoid_algebra.algebra
/-
TODO: add `add_monoid_algebra.is_scalar_tower` so that we have
instance [comm_semiring S₂] [has_scalar R S₁] [semimodule R S₂] [semimodule S₁ S₂]
  [is_scalar_tower R S₁ S₂] : is_scalar_tower R S₁ (mv_polynomial σ S₂) :=
  add_monoid_algebra.is_scalar_tower
-/
```